domain/annotation/state/state.go

```diff
@@ -0,0 +1,292 @@
+	if err != nil {
+		return errors.Annotatef(err, "preparing get annotations query for ID: %q", ID.Name)
+	}
+	deleteAnnotationsStmt, err = sqlair.Prepare(deleteAnnotationsQuery, Annotation{}, sqlair.M{})
+	if err != nil {
+		return errors.Annotatef(err, "preparing get annotations query for ID: %q", ID.Name)
+	}
+
+	// Running transactions using sqlair statements
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		// If it's for a model, go ahead and run the queries (no uuid needed)
+		if kind == annotations.KindModel {
+			// Unset the annotations
+			if err := tx.Query(ctx, deleteAnnotationsStmt).Run(); err != nil {
+				return errors.Annotatef(err, "unsetting annotations for ID: %s", ID.Name)
+			}
+
+			// Insert annotations
+			for key, value := range toUpsert {
+				if err := tx.Query(ctx, setAnnotationsStmt, sqlair.M{
+					"key":   key,
+					"value": value,
+				}).Run(); err != nil {
+					return errors.Annotatef(err, "setting annotations for ID: %s", ID.Name)
+				}
+			}
+			return nil
+		}
+
+		// For non-model kinds, we need to find the uuid of ID first
+		// Looking up the UUID for ID
+		result := sqlair.M{}
+		err = tx.Query(ctx, kindQueryStmt, kindQueryParam).Get(result)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(err, "looking up UUID for ID: %s", ID.Name)
+		}
+
+		if len(result) == 0 {
+			return errors.Annotatef(err, "unable to find UUID for ID: %q", ID.Name)
+		}
+		uuid := result["uuid"].(string)
+
+		// Unset the annotations
+		if err := tx.Query(ctx, deleteAnnotationsStmt, sqlair.M{
+			"uuid": uuid,
+		}).Run(); err != nil {
+			return errors.Annotatef(err, "unsetting annotations for ID: %s", ID.Name)
+		}
+
+		// Insert annotations
+		for key, value := range toUpsert {
+			if err := tx.Query(ctx, setAnnotationsStmt, sqlair.M{
+				"uuid":  uuid,
+				"key":   key,
+				"value": value,
+			}).Run(); err != nil {
+				return errors.Annotatef(err, "setting annotations for ID: %s", ID.Name)
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return errors.Annotatef(err, "setting annotations for ID: %q", ID.Name)
+	}
+
+	return nil
+}
+
+// uuidQueryForID is a helper that generates a query and parameters for getting the uuid for a given ID
+// We keep different fields to reference different IDs in separate tables, as follows:
+// machine: TABLE machine, reference field: machine_id
+// unit: TABLE unit, reference field: unit_id
+// application: TABLE application, reference field: name
+// storage_instance: TABLE storage_instance, reference field: name
+// charm: TABLE charm, reference field: url
+func uuidQueryForID(ID annotations.ID) (string, sqlair.M, error) {
+	kind := ID.Kind
+	name := ID.Name
+
+	var query string
+
+	if kind == "machine" || kind == "unit" {
+		// Use field <entity>_id (e.g. unit_id) for machines and units
+		query = fmt.Sprintf(`SELECT &M.uuid FROM %[1]s WHERE %[1]s_id = $M.entity_id`, kind)
```

Comment:
Try to remove the sprintf if possible and use a `?`. This can lead to SQL injection if we're not careful.